### PR TITLE
Streamk v0.2

### PR DESCRIPTION
**Pull request by xiaohuguo2023**
streamk v0.2:

- new streamk tuning script to reduce compiling and profiling time

- use load/store cache modifier to reimplement spinning lock

- add CI test for streamk-kernel

- able to use streampipelineV2

### CODE REVIEW

File: python/perf-kernels/streamk/streamk_kernel.py
Comment (neoblizz) [inline python/perf-kernels/streamk/streamk_kernel.py:1]: Note, for gfx90a the load/stores with `cache_modifiers` do not work. Documented here: https://github.com/ROCm/triton-internal/issues/311

File: python/perf-kernels/streamk/streamk_kernel.py [hunk 1]
```diff
@@ -166,25 +146,34 @@ def streamk_gemm(
...
-        if start_iter == tile_iter:
-            tile_iter_end = tile_iter + iters_per_tile
+
+        if start_iter != tile_iter:
+            rm1 = tl.arange(0, BLOCK_SIZE_M)
+            rn1 = tl.arange(0, BLOCK_SIZE_N)
+            rm1 = tl.max_contiguous(tl.multiple_of(rm1, BLOCK_SIZE_M), BLOCK_SIZE_M)
+            rn1 = tl.max_contiguous(tl.multiple_of(rn1, BLOCK_SIZE_N), BLOCK_SIZE_N)
+            P_ = P + pid * BLOCK_SIZE_M * BLOCK_SIZE_N + rm1[:, None] * BLOCK_SIZE_N + rn1[None, :]
+            tl.store(P_, acc, cache_modifier=".wt")
```
Comment (neoblizz) [inline python/perf-kernels/streamk/streamk_kernel.py:156]: Note, for gfx90a the load/stores with `cache_modifiers` do not work. Documented here: https://github.com/ROCm/triton-internal/issues/311

File: python/perf-kernels/streamk/streamk_kernel.py [hunk 2]
```diff
@@ -166,25 +146,34 @@ def streamk_gemm(
...
+            tl.store(locks + pid, 1, cache_modifier=".wt")
+        else:
             next_pid = pid + 1
+            tile_iter_end = tile_iter + iters_per_tile
             end = end_iter
-            while (end < tile_iter_end and next_pid < num_cus):
-                # todo: try use tl.load once cache modifier landed upstream
-                while tl.atomic_cas(locks + next_pid, 1, 1) != 1:
+            while (end < tile_iter_end and next_pid < NUM_SMS):
+                while tl.load(locks + next_pid, cache_modifier=".cv", volatile=True) != 1:
```
Comment (neoblizz) [inline python/perf-kernels/streamk/streamk_kernel.py:163]: This also does not work in `gfx90a`: https://github.com/ROCm/triton-internal/issues/311

File: python/perf-kernels/streamk/streamk_kernel.py [hunk 3]
```diff
@@ -2,121 +2,100 @@
...
     BLOCK_SIZE_N: tl.constexpr,
     BLOCK_SIZE_K: tl.constexpr,
     GROUP_SIZE_M: tl.constexpr,
+    NUM_SMS: tl.constexpr,
+    BIAS: tl.constexpr,
     EVEN_K: tl.constexpr,
 ):
     pid = tl.program_id(0)
-    pid = get_new_pid(pid, num_cus)
+    pid = (pid % 8) * (NUM_SMS // 8) + (pid // 8)
```
Comment (neoblizz) [inline python/perf-kernels/streamk/streamk_kernel.py:32]: This is not needed for anything but `gfx942`, so we will actually remove this if the arch was `gfx90a`.

File: python/perf-kernels/streamk/tune_streamk.py [hunk 4]
```diff
@@ -535,56 +383,52 @@ def matmul(a, b, c, P, locks, num_cus, block_m, block_n, block_k, group_m, num_w
...
+        bias_fp16 = bias.squeeze()
     # Allocates output.
-    print(f"{block_k}")
-    EVEN_K = K % block_k == 0
     c = torch.zeros((M, N), device=a.device, dtype=tl_to_torch_types[name_to_tl_types[dtype_c]])
-    locks = torch.zeros((num_cus, ), device="cuda", dtype=torch.int32)
-    P = torch.zeros((num_cus, block_m * block_n), device="cuda", dtype=torch.float32)
-    triton_output = matmul(a, b, c, P, locks, num_cus, block_m, block_n, block_k, group_m, num_warps, num_stages,
-                           waves_per_eu, mfmaInstrSize, kpack, EVEN_K)
+    locks = torch.zeros((num_sms, ), device="cuda", dtype=torch.int32)
```
Comment (neoblizz) [inline python/perf-kernels/streamk/tune_streamk.py:415]: `locks` can be less than `int32` type, we only need 1 byte: `uint8` should work.

**Review by neoblizz — COMMENTED** [review]
I put in some comments.

**Comment by xiaohuguo2023** [comment]
let's close this PR as there are too many difference with new main_perf, it's not safe to merge anymore. I have created a new ]PR]( https://github.com/ROCm/triton/pull/652) for v0.2

**Pull request closed without merging**